### PR TITLE
JAVA-2527: Allow AllNodesFailedException to accept more than one error per node

### PR DESCRIPTION
**Pull request by adutra**
Note: this PR fixes the core issue and exposes a new getter returning a `Map<Node, List<Throwable>>` structure containing all errors (it also deprecates the old `getErrors` method).

It does not add suppressed exceptions as suggested in JAVA-2528.

### CODE REVIEW

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java [hunk 1]
```diff
@@ -17,87 +17,150 @@
...
       @NonNull String message,
       @Nullable ExecutionInfo executionInfo,
       @NonNull Map<Node, Throwable> errors) {
+    this(message, executionInfo, groupByNode(errors.entrySet()));
+  }
+
+  protected AllNodesFailedException(
+      @NonNull String message,
+      @Nullable ExecutionInfo executionInfo,
+      @NonNull Iterable<Entry<Node, List<Throwable>>> errors) {
```
Comment (adutra) [inline core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java:73]: Note that this can't be a `Map<Node, List<Throwable>>` because the erased type `Map` would clash with the existing constructor signature. This unfortunately makes the fix not as straightforward as it would seem.

File: integration-tests/src/test/java/com/datastax/oss/driver/core/AllNodesFailedIT.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map.Entry;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+@Category(ParallelizableTests.class)
+public class AllNodesFailedIT {
```
Comment (adutra) [inline integration-tests/src/test/java/com/datastax/oss/driver/core/AllNodesFailedIT.java:47]: I tried to create a unit test first using the existing test harness, but spent too long trying to get it to work, so I went for an IT instead.

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java [hunk 3]
```diff
@@ -17,87 +17,150 @@
...
+    while (iterator.hasNext()) {
+      Entry<Node, List<Throwable>> entry = iterator.next();
+      details.append(entry.getKey()).append(": ").append(entry.getValue());
+      if (iterator.hasNext()) {
+        details.append(", ");
+      }
+    }
     return String.format(
-        baseMessage + " (showing first %d, use getErrors() for more: %s)", limit, details);
+        baseMessage + " (showing first %d, use getAllErrors() for more): %s", limit, details);
```
Comment (adutra) [inline core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java]: Nit: I allowed myself to move the closing parenthesis before the errors list, which can be very long; otherwise, it seems that the opening parenthesis is unbalanced.

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java [hunk 4]
```diff
@@ -17,87 +17,150 @@
...
     return String.format(
-        baseMessage + " (showing first %d, use getErrors() for more: %s)", limit, details);
+        "%s (showing first %d, use getAllErrors() for more): %s", baseMessage, limit, details);
   }
 
-  /** The details of the individual error on each node. */
+  /**
+   * An immutable map containing the first error on each tried node.
+   *
+   * @deprecated Use {@link #getAllErrors()} instead.
```
Comment (olim7t) [inline core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java:106]: Maybe we should also indicate that this version will only report the first error of each node.

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java [hunk 5]
```diff
@@ -17,87 +17,150 @@
...
   }
 
-  private static String buildMessage(String baseMessage, Map<Node, Throwable> errors) {
+  private static String buildMessage(
+      String baseMessage, Collection<Entry<Node, List<Throwable>>> errors) {
     int limit = Math.min(errors.size(), 3);
-    String details =
-        Joiner.on(", ").withKeyValueSeparator(": ").join(Iterables.limit(errors.entrySet(), limit));
-
+    Iterator<Entry<Node, List<Throwable>>> iterator = errors.iterator();
```
Comment (olim7t) [inline core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java]: We've lost the limit in the refactoring. You can use `Iterables.limit()` on this line.

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java [hunk 6]
```diff
@@ -17,87 +17,150 @@
...
+    while (iterator.hasNext()) {
+      Entry<Node, List<Throwable>> entry = iterator.next();
+      details.append(entry.getKey()).append(": ").append(entry.getValue());
+      if (iterator.hasNext()) {
+        details.append(", ");
+      }
+    }
     return String.format(
-        baseMessage + " (showing first %d, use getErrors() for more: %s)", limit, details);
+        "%s (showing first %d, use getAllErrors() for more): %s", baseMessage, limit, details);
```
Comment (olim7t) [inline core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java]: Maybe reword as `showing first %d nodes`, it's what it will do if you apply my other suggestion a few lines above.

**Review by olim7t — APPROVED** [review]
LGTM 👍 

Minor formatting suggestions inline.

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java [hunk 7]
```diff
@@ -17,87 +17,150 @@
...
   public AllNodesFailedException reword(String newMessage) {
     return new AllNodesFailedException(
-        buildMessage(newMessage, errors), getExecutionInfo(), errors);
+        buildMessage(newMessage, errors.entrySet()), getExecutionInfo(), errors.entrySet());
+  }
+
+  private static Set<Entry<Node, List<Throwable>>> groupByNode(
+      Iterable<Entry<Node, Throwable>> errors) {
+    // no need for immutable collections here
+    Map<Node, List<Throwable>> map = new LinkedHashMap<>();
```
Comment (tomekl007) [inline core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java:145]: Isn't it better to just use guaga `MultiMap`?

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java [hunk 8]
```diff
@@ -17,87 +17,150 @@
...
+          });
+    }
+    return map.entrySet();
+  }
+
+  private static Map<Node, List<Throwable>> toDeepImmutableMap(
+      Iterable<Entry<Node, List<Throwable>>> errors) {
+    ImmutableMap.Builder<Node, List<Throwable>> builder = ImmutableMap.builder();
+    for (Entry<Node, List<Throwable>> entry : errors) {
+      builder.put(entry.getKey(), ImmutableList.copyOf(entry.getValue()));
```
Comment (tomekl007) [inline core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java:170]: If that is a deep copy, shouldn't we make a safe copy of `entry.getKey()`? Or is the Node returned by `getKey` immutable?

File: core/src/main/java/com/datastax/oss/driver/internal/core/control/ControlConnection.java [hunk 9]
```diff
@@ -573,20 +576,21 @@ private void forceClose() {
...
       if (errors.size() == 0) {
         return false;
       }
-      for (Throwable nodeError : errors) {
-        if (!(nodeError instanceof AuthenticationException)) {
-          authFailure = false;
-          break;
+      for (List<Throwable> nodeErrors : errors) {
+        for (Throwable nodeError : nodeErrors) {
+          if (!(nodeError instanceof AuthenticationException)) {
```
Comment (tomekl007) [inline core/src/main/java/com/datastax/oss/driver/internal/core/control/ControlConnection.java:587]: is this refactoring correct? Previously if any of the errors were `AuthenticationException` then this method returned `true`. Now if any of the errors is NOT `AuthenticationException` then this method returns `false`.

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java (hunk 4 above)
Comment (adutra) [reply core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java:106]: That's already stated on line 103, do you think it's worth repeating it here (line 105)?

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java (hunk 7 above)
Comment (adutra) [reply core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java:145]: Not possible, we cannot expose Guava in the API (it's shaded).

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java (hunk 8 above)
Comment (adutra) [reply core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java:170]: `DefaultNode` is mutable, but I don't want to change it in this PR because it was already mutable in the previous implementation of this class. I'm only making the values immutable.

File: core/src/main/java/com/datastax/oss/driver/internal/core/control/ControlConnection.java (hunk 9 above)
Comment (adutra) [reply core/src/main/java/com/datastax/oss/driver/internal/core/control/ControlConnection.java:587]: I don't agree, both algorithms look equivalent to me and both return `false` as soon as any error is NOT `AuthenticationException`.

Comment (tomekl007) [reply core/src/main/java/com/datastax/oss/driver/internal/core/control/ControlConnection.java:587]: ah, ok: there is a `break` there, thx for the clarification

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java (hunk 5 above)
Comment (adutra) [reply core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java]: The only thing that bothers me is that the signature of this method, at least in theory, cannot guarantee that exactly N distinct nodes will be printed, since duplicate entries could happen (not in the current implementation, but if someone extends this class, then it's not impossible). I will try to change the signature to accept a map of nodes instead, just to be on the safe side.

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java (hunk 6 above)
Comment (adutra) [reply core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java]: Done, but I also changed the signature to accept a map of nodes instead. All in all, it looks a bit more straightforward this way than before imo.

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java (hunk 4 above)
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java:106]: I don't know how I managed to miss that 🤦‍♂
Forget this comment then.

File: core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java (hunk 8 above)
Comment (olim7t) [reply core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java:170]: `DefaultNode` is mutable but it uses reference equality. Changing the mutable fields will never affect how two instances compare.
